lesson4/hometask.rb

```diff
@@ -0,0 +1,116 @@
+def test(klass)
+  lazy_hash = klass.new
+  lazy_hash[:proc] = proc { Time.now }
+  lazy_hash[:lambda] = lambda { Time.now }
+  lazy_hash[:raw] = Time.now
+
+  result = true
+
+  result = result && lazy_hash[:proc].is_a?(Proc)
+  result = result && lazy_hash[:lambda].is_a?(Proc)
+  result = result && lazy_hash[:raw].is_a?(Time)
+
+  result = result && lazy_hash.lazy?(:proc)
+  result = result && lazy_hash.lazy?(:lambda)
+  result = result && !lazy_hash.lazy?(:raw)
+
+  result = result && lazy_hash.respond_to?(:proc)
+  result = result && lazy_hash.respond_to?(:lambda)
+  result = result && lazy_hash.respond_to?(:raw)
+
+  result = result && Time.now < lazy_hash.proc
+  result = result && Time.now < lazy_hash.lambda
+  result = result && Time.now > lazy_hash.raw
+
+  result = result && lazy_hash.proc.is_a?(Time)
+  result = result && lazy_hash.lambda.is_a?(Time)
+  result = result && lazy_hash.raw.is_a?(Time)
+
+  result = result && Time.now > lazy_hash.proc
+  result = result && Time.now > lazy_hash.lambda
+  result = result && Time.now > lazy_hash.raw
+
+  result = result && lazy_hash[:proc] == lazy_hash.proc
+  result = result && lazy_hash[:lambda] == lazy_hash.lambda
+  result = result && lazy_hash[:raw] == lazy_hash.raw
+
+  result = result && !lazy_hash.lazy?(:proc)
+  result = result && !lazy_hash.lazy?(:lambda)
+  result = result && !lazy_hash.lazy?(:raw)
+
+  puts result
+end
+
+
+#==============================================
+# SimpleDelegator usage and no method missing
+#==============================================
+
+require 'delegate'
+
+class LazyHashDelegator < SimpleDelegator
+  def initialize(initial_hash = {})
+    self.__setobj__(initial_hash)
```

Comment:
It could be redundant if your test allows as to specify `hash` as decorated object directly, without using default value)